[docs][data grid] Add documentation for the list view feature (#15344)
mui/mui-x: docs/data/data-grid/list-view/list-view.md

- ✅ [Row pinning](/x/react-data-grid/row-pinning/) [<span class="plan-pro"></span>](/x/introduction/licensing/#pro-plan 'Pro plan')
- ✅ [Pagination](/x/react-data-grid/pagination/)
- ✅ [Row selection](/x/react-data-grid/row-selection/)
- ✅ [Cell selection](/x/react-data-grid/cell-selection/) [<span class="plan-premium"></span>](/x/introduction/licensing/#premium-plan 'Premium plan')

:::warning
Features not listed may not work as expected, or may not work at all.

If you're using a feature that's listed above and it's not working as expected, please [open a bug report](https://github.com/mui/mui-x/issues/new?assignees=&labels=status%3A+waiting+for+maintainer%2Cbug+%F0%9F%90%9B&projects=&template=1.bug.yml).

If you need to use list view with any other features, please [open a feature request](https://github.com/mui/mui-x/issues/new?assignees=&labels=status%3A+waiting+for+maintainer%2Cnew+feature&projects=&template=2.feature.yml).
:::

## Selectors

{{"component": "modules/components/SelectorsDocs.js", "category": "List View"}}

## API

- [DataGrid](/x/api/data-grid/data-grid/)
- [DataGridPro](/x/api/data-grid/data-grid-pro/)

> - [DataGridPremium](/x/api/data-grid/data-grid-premium/)

Reviewer: Since we have the interface documented, does it make sense to include it here?
```suggestion
- [DataGridPremium](/x/api/data-grid/data-grid-premium/)
- [GridListColDef](/x/api/data-grid/grid-list-col-def/)
```

In order to display data in a list view, a `unstable_listColumn` prop must be provided with a `renderCell` function.

```tsx
function ListViewCell(params: GridRenderCellParams) {
  return <>{params.row.id}</>;
}

const listColDef: GridListColDef = {
  field: 'listColumn',
  renderCell: ListViewCell,
};
```

{{"demo": "ListView.js", "bg": true}}

## Enable with a media query

> The `useMediaQuery` hook from `@mui/material` can be used to enable the list view feature at a specified breakpoint.

Reviewer 2: passive –> active voice

```suggestion
Use the `useMediaQuery` hook from `@mui/material` to enable the list view feature at a specified breakpoint.
```

```tsx
import * as React from 'react';
import useMediaQuery from '@mui/material/useMediaQuery';

function App() {
  const theme = useTheme();
  const isListView = useMediaQuery(theme.breakpoints.down('md'));

  return <DataGrid unstable_listView={isListView} />;
}
```

## Editable rows

> The [editing](/x/react-data-grid/editing/) feature is not supported in list view, however, it is possible to build an editing experience from within your custom cell renderer.

Reviewer 2: - best practice: longer anchor text
- "but" is usually always better than "however" (it's more concise and less formal)
- point out that the demo that follows shows how to do this

```suggestion
The [editing feature](/x/react-data-grid/editing/) is not supported in list view, but it is possible to build an editing experience from within your custom cell renderer, as shown below.
```

{{"demo": "ListViewEdit.js", "bg": true}}

## Field visibility

> Similar to the default grid view, field visibilty can be toggled in list view.
> 
> In the list view column's `renderCell` function, data can be conditionally rendered based on [column visibility](/x/react-data-grid/column-visibility/).

Reviewer 2: ```suggestion
Field visibility can be toggled in list view, similarly to the default grid view.
In the list view column's `renderCell` function, data can be conditionally rendered based on [column visibility](/x/react-data-grid/column-visibility/), as shown in the code snippet below.
```

```tsx
import {
  useGridSelector,
  useGridApiContext,
  gridColumnVisibilityModelSelector,
} from '@mui/x-data-grid';

function ListViewCell(params: GridRenderCellParams) {
  const apiRef = useGridApiContext();
  const columnVisibilityModel = useGridSelector(
    apiRef,
    gridColumnVisibilityModelSelector,
  );
  const showCreatedAt = columnVisibilityModel.createdBy !== false;

  return (
    <>
      <span>{params.row.id}</span>
      {showCreatedAt && (
        <time datetime={params.row.createdAt}>
          {formatDate(params.row.createdAt)}
        </time>
      )}
    </>
  );
}
```

## Advanced usage

The list view feature can be combined with [custom subcomponents](/x/react-data-grid/components/) to provide an improved user experience on small screens.

{{"demo": "ListViewAdvanced.js", "iframe": true, "maxWidth": 360, "height": 600}}

> ## Feature support

Author: How do we feel about this section?

Reviewer 2: I'm cool with it. Maybe "Feature interoperability" would be more accurate since we're specifically describing how this feature works with other features. But that's also a pretty gnarly word to throw at folks 😅 especially non-native English speakers.

Author: That is gnarly 🤙.

What do others think? @mui/xgrid do we have a similar section anywhere else?

Reviewer: For me, `Feature compatibility` would be more understandable.

Another possibility could be to convert the section to a `::success` callout following the pattern for `::warning` callout in https://mui.com/x/react-data-grid/column-spanning/, I'd prefer keeping it as a separate section though.

Author: I like "Feature compatability", makes more sense than "support" when comparing to other features 👍

Author: I like "Feature compatibility", makes more sense than "support" when comparing to other features 👍

The list view feature can be used in combination with the following features:

- ✅ [Sorting](/x/react-data-grid/sorting/)
- ✅ [Filtering](/x/react-data-grid/filtering/)
- ✅ [Multi filters](/x/react-data-grid/filtering/multi-filters/) [<span class="plan-pro"></span>](/x/introduction/licensing/#pro-plan 'Pro plan')
- ✅ [Row pinning](/x/react-data-grid/row-pinning/) [<span class="plan-pro"></span>](/x/introduction/licensing/#pro-plan 'Pro plan')
- ✅ [Pagination](/x/react-data-grid/pagination/)
- ✅ [Row selection](/x/react-data-grid/row-selection/)
- ✅ [Cell selection](/x/react-data-grid/cell-selection/) [<span class="plan-premium"></span>](/x/introduction/licensing/#premium-plan 'Premium plan')

:::warning
Features not listed may not work as expected, or may not work at all.

> If you are using a feature that is listed and it is not working as expected, please [open a bug report](https://github.com/mui/mui-x/issues/new?assignees=&labels=status%3A+waiting+for+maintainer%2Cbug+%F0%9F%90%9B&projects=&template=1.bug.yml).

Reviewer 2: Don't be afraid of contractions—they're great 😁

```suggestion
If you're using a feature that's listed above and it's not working as expected, please [open a bug report](https://github.com/mui/mui-x/issues/new?assignees=&labels=status%3A+waiting+for+maintainer%2Cbug+%F0%9F%90%9B&projects=&template=1.bug.yml).
```

In order to display data in a list view, a `unstable_listColumn` prop must be provided with a `renderCell` function.

> ```tsx
> function ListViewCell(params: GridRenderCellParams) {
>   return <>{params.row.id}</>;
> }
> 
> const listColDef: GridListColDef = {
>   field: 'listColumn',
>   renderCell: ListViewCell,
> };
> ```

Reviewer: To connect the snippet with the above paragraphs:
```suggestion
```tsx
function ListViewCell(params: GridRenderCellParams) {
  return <>{params.row.id}</>;
}

const listColDef: GridListColDef = {
  field: 'listColumn',
  renderCell: ListViewCell,
};

<DataGridPro
  unstable_listColumn={listColDef}
  unstable_listView={true}
/>
```
```

{{"demo": "ListView.js", "bg": true}}

## Enable with a media query

Use the `useMediaQuery` hook from `@mui/material` to enable the list view feature at a specified breakpoint.

The demo below automatically switches to a list layout when the viewport width is below the `md` breakpoint.

{{"demo": "ListViewMediaQuery.js", "bg": "inline"}}

## Editable rows

The [editing feature](/x/react-data-grid/editing/) is not supported in list view, but it is possible to build an editing experience from within your custom cell renderer, as shown below.

{{"demo": "ListViewEdit.js", "bg": true}}

## Field visibility

> Field visibility can be toggled in list view, similarly to the default grid view.

Reviewer: I'm wondering if "Advanced usage" section also covers this aspect and if we could skip this section (for simplicity) 🤔 
What do you think?

Author: Yeah, in hindsight I don't think it is necessary to have a section for this.
